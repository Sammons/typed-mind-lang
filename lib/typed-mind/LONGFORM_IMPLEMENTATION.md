# TypedMind Longform Syntax Implementation Guide

This document describes the implementation details for TypedMind's dual syntax support. For the complete grammar reference, see [grammar.md](./grammar.md).

TypedMind supports both shortform (for LLMs) and longform (for humans) syntax. Both compile to the same AST.

## Syntax Comparison

### Program Declaration
```tmd
# Shortform
TodoApp -> AppEntry v1.0.0

# Longform
program TodoApp {
  entry: AppEntry
  version: "1.0.0"
}
```

### File Declaration
```tmd
# Shortform
AppEntry @ src/index.ts:
  <- [Express, Database]
  -> [startServer]

# Longform
file AppEntry {
  path: "src/index.ts"
  imports: [Express, Database]
  exports: [startServer]
}
```

### Function Declaration
```tmd
# Shortform
createUser :: (data: UserDTO) => Promise<User>
  "Creates a new user"
  <- UserDTO
  -> User
  ~> [validate, Database.save]
  +> [UserList]
  $< [DATABASE_URL, API_KEY]

# Longform
function createUser {
  signature: "(data: UserDTO) => Promise<User>"
  description: "Creates a new user"
  input: UserDTO
  output: User
  calls: [validate, Database.save]
  affects: [UserList]
  consumes: [DATABASE_URL, API_KEY]
}
```

### Class Declaration
```tmd
# Shortform
UserService <: BaseService, IService
  => [create, read, update, delete]

# Longform
class UserService {
  extends: BaseService
  implements: [IService]
  methods: [create, read, update, delete]
}
```

### DTO Declaration
```tmd
# Shortform
UserDTO % "User data transfer object"
  - name: string "User's full name"
  - email: string "Email address"
  - age: number "Age in years" (optional)

# Longform
dto UserDTO {
  description: "User data transfer object"
  fields: {
    name: {
      type: "string"
      description: "User's full name"
    }
    email: {
      type: "string"
      description: "Email address"
    }
    age: {
      type: "number"
      description: "Age in years"
      optional: true
    }
  }
}
```

### UI Component Declaration
```tmd
# Shortform
UserProfile & "User profile component"
  < [Dashboard]
  > [Avatar, UserInfo]
  <+ [updateProfile, refreshData]

# Longform
component UserProfile {
  description: "User profile component"
  containedBy: [Dashboard]
  contains: [Avatar, UserInfo]
  affectedBy: [updateProfile, refreshData]
}

# Root component shortform
App &! "Root application component"

# Root component longform
component App {
  description: "Root application component"
  root: true
}
```

### Asset Declaration
```tmd
# Shortform
Logo ~ "Company logo SVG"
IndexHTML ~ "Main HTML file"
  >> ClientApp

# Longform
asset Logo {
  description: "Company logo SVG"
}

asset IndexHTML {
  description: "Main HTML file"
  containsProgram: ClientApp
}
```

### Constants Declaration
```tmd
# Shortform
Config ! src/config.ts : ConfigSchema

# Longform
constants Config {
  path: "src/config.ts"
  schema: ConfigSchema
}
```

### RunParameter Declaration
```tmd
# Shortform
DATABASE_URL $env "PostgreSQL connection" (required)
API_KEY $env "API authentication key"
  = "dev-key-123"
LAMBDA_ROLE $iam "Execution role"
NODE_VERSION $runtime "Node.js version"
  = "20.x"
MAX_CONNECTIONS $config "Connection pool size"
  = "100"

# Longform
parameter DATABASE_URL {
  type: "env"
  description: "PostgreSQL connection"
  required: true
}

parameter API_KEY {
  type: "env"
  description: "API authentication key"
  default: "dev-key-123"
}

parameter LAMBDA_ROLE {
  type: "iam"
  description: "Execution role"
}

parameter NODE_VERSION {
  type: "runtime"
  description: "Node.js version"
  default: "20.x"
}

parameter MAX_CONNECTIONS {
  type: "config"
  description: "Connection pool size"
  default: "100"
}
```

### Import Declaration
```tmd
# Shortform
@import "./shared/auth.tmd" as Auth
@import "./utils.tmd"

# Longform
import "./shared/auth.tmd" as Auth
import "./utils.tmd"
```

## Implementation Notes

1. **Lexer Changes**: Add tokens for `program`, `file`, `function`, `class`, `dto`, `component`, `asset`, `constants`, `parameter`, `import`, curly braces `{` `}`, and colon `:`

2. **Parser Changes**: 
   - Check for longform keywords at the start of entity declarations
   - Parse block-style syntax with key-value pairs
   - Maintain backward compatibility with shortform

3. **AST**: No changes needed - both syntaxes produce the same AST

4. **Validation**: No changes needed - operates on the AST

5. **VS Code Extension**: Update syntax highlighting to support both forms

## Benefits

- **Human-readable**: Longform is self-documenting with clear property names
- **LLM-efficient**: Shortform remains terse for AI generation
- **Flexible**: Users can mix both syntaxes in the same file
- **Migration-friendly**: Existing shortform files continue to work