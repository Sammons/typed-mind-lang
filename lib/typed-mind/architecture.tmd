# Programs
IndexApp -> IndexFile v1.0.0

# Dependencies
Path ^ "Path manipulation utilities"
Fs ^ "File system operations"

# Files
IndexFile @ lib/typed-mind/src/index.ts:
  <- [DSLParser, DSLValidator, ErrorFormatter, ImportResolver, ValidationResult, ProgramGraph, AnyEntity, ValidationError, Path]
  -> [DSLChecker]

# ClassFiles (Services/Controllers)
DSLParser #: lib/typed-mind/src/parser.ts
  <- [AnyEntity, ProgramEntity, FileEntity, FunctionEntity, ClassEntity, ClassFileEntity, ConstantsEntity, DTOEntity, DTOField, AssetEntity, UIComponentEntity, RunParameterEntity, DependencyEntity, Position, ImportStatement, LongformParser, ENTITY_PATTERNS, CONTINUATION_PATTERNS, GENERAL_PATTERNS, GrammarValidator]
  => [parse]
  -> [ParseResult]
DSLValidator #: lib/typed-mind/src/validator.ts
  <- [AnyEntity, ClassEntity, FunctionEntity, UIComponentEntity, AssetEntity, RunParameterEntity, ConstantsEntity, DependencyEntity, DTOEntity, ProgramEntity, ValidationError, ValidationResult, EntityType, ReferenceType, ParseResult]
  => [validate]
  -> [ValidatorOptions]
ErrorFormatter #: lib/typed-mind/src/formatter.ts
  <- [ValidationError]
  => [format, formatAll]
ImportResolver #: lib/typed-mind/src/import-resolver.ts
  <- [Fs, Path, ImportStatement, AnyEntity, ValidationError, ParseResult, DSLParser]
  => [resolveImports]
  -> [ResolvedImport]
LongformParser #: lib/typed-mind/src/longform-parser.ts
  <- [AnyEntity, ProgramEntity, FileEntity, FunctionEntity, ClassEntity, ConstantsEntity, DTOEntity, DTOField, AssetEntity, UIComponentEntity, RunParameterEntity, DependencyEntity, Position]
  => [parseLongform, getConsumedLines]
GrammarValidator #: lib/typed-mind/src/grammar-validator.ts
  <- [AnyEntity, EntityType, ENTITY_TYPE_NAMES, EntityTypeName]
  => [validateEntity, validateEntities, formatErrors]
  -> [GrammarValidationError, GrammarValidationResult]

# Classes
DSLChecker <:
  => [check, parse]

# DTOs
ParseResult %
  - entities: Map<string, AnyEntity>
  - imports: ImportStatement[]
  - namingConflicts?: Array<{
    name: string;
    existingEntity: AnyEntity;
    newEntity: AnyEntity;
  }>
  - grammarErrors?: Array<{
    entity: string;
    type: string;
    field: string;
    message: string;
  }>
ValidatorOptions %
  - skipOrphanCheck?: boolean
ResolvedImport %
  - import: ImportStatement
  - entities: Map<string, AnyEntity>
  - imports: ImportStatement[]
  - errors?: ValidationError[]
LongformBlock %
  - type: string
  - name: string
  - properties: Map<string, any>
  - position: Position
  - raw: string
GrammarValidationError %
  - entity: string
  - type: EntityType
  - field: string
  - expected: string
  - actual: string | undefined
  - message: string
GrammarValidationResult %
  - valid: boolean
  - errors: GrammarValidationError[]
Position %
  - line: number
  - column: number
Reference %
  - from: string
  - type: ReferenceType
  - fromType?: EntityType
Entity %
  - name: string
  - type: EntityType
  - position: Position
  - raw: string
  - comment?: string
  - referencedBy?: Reference[]
ProgramEntity %
  - type: string
  - entry: string
  - version?: string
  - purpose?: string
  - exports?: string[]
FileEntity %
  - type: string
  - path: string
  - imports: string[]
  - exports: string[]
  - purpose?: string
FunctionEntity %
  - type: string
  - container?: string
  - signature: string
  - description?: string
  - calls: string[]
  - input?: string
  - output?: string
  - affects?: string[]
  - consumes?: string[]
ClassEntity %
  - type: string
  - container?: string
  - path?: string
  - extends?: string
  - implements: string[]
  - methods: string[]
  - imports?: string[]
  - purpose?: string
ClassFileEntity %
  - type: string
  - path: string
  - extends?: string
  - implements: string[]
  - methods: string[]
  - imports: string[]
  - exports: string[]
  - purpose?: string
ConstantsEntity %
  - type: string
  - path: string
  - schema?: string
  - purpose?: string
DTOField %
  - name: string
  - type: string
  - description?: string
  - optional?: boolean
DTOEntity %
  - type: string
  - purpose?: string
  - fields: DTOField[]
AssetEntity %
  - type: string
  - description: string
  - containsProgram?: string
UIComponentEntity %
  - type: string
  - purpose: string
  - root?: boolean
  - contains?: string[]
  - containedBy?: string[]
  - affectedBy?: string[]
RunParameterEntity %
  - type: string
  - paramType: string
  - description: string
  - defaultValue?: string
  - required?: boolean
  - consumedBy?: string[]
DependencyEntity %
  - type: string
  - purpose: string
  - version?: string
  - importedBy?: string[]
  - exports?: string[]
ValidationError %
  - position: Position
  - message: string
  - severity: string
  - suggestion?: string
ValidationResult %
  - valid: boolean
  - errors: ValidationError[]
ImportStatement %
  - path: string
  - alias?: string
  - position: Position
ProgramGraph %
  - entities: Map<string, AnyEntity>
  - dependencies: Map<string, string[]>
  - imports?: ImportStatement[]

# Constants
EntityType ! src/types.ts
ReferenceType ! src/types.ts
AnyEntity ! src/types.ts
EntityTypeName ! src/types.ts
ENTITY_PATTERNS ! lib/typed-mind/src/parser-patterns.ts
CONTINUATION_PATTERNS ! lib/typed-mind/src/parser-patterns.ts
GENERAL_PATTERNS ! lib/typed-mind/src/parser-patterns.ts
ENTITY_TYPE_NAMES ! lib/typed-mind/src/parser-patterns.ts
PATTERN_DESCRIPTIONS ! lib/typed-mind/src/parser-patterns.ts
