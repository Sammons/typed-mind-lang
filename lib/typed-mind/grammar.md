# TypedMind DSL Grammar Reference

This document is auto-generated from the parser patterns.

## Note from Author
TypedMind is meant to be a DSL to represent a variety of programs and
force AI to create a cohesive program architecture with a relatively token efficient syntax.

Entities link bidirectionally, so for example it is not enough to declare a function,
the file must also be declared. The function must be exported by a file. And the function must be 
consumed by another entity to avoid dead code. The TypeMind checker will validate these scenarios.

## Table of Contents

1. [Entity Types](#entity-types)
2. [Entity Patterns](#entity-patterns)
3. [Continuation Patterns](#continuation-patterns)
4. [General Patterns](#general-patterns)
5. [Examples](#examples)

## Entity Types

TypedMind supports the following entity types:

| Entity Type | Description |
|------------|-------------|
| Program | Defines an application entry point |
| File | Defines a source code file |
| Function | Defines a function with its type signature |
| Class | Defines a class with inheritance |
| ClassFile | Combines class and file definitions in one entity - perfect for services, controllers, and modules |
| Constants | Defines a constants/configuration file |
| DTO | Defines a Data Transfer Object for data structures (config, parameters, serialization) - NO function fields allowed |
| Asset | Defines a static asset |
| UIComponent | Defines a UI component (&! for root) |
| RunParameter | Defines a runtime parameter |
| Dependency | Defines an external dependency |

## Entity Patterns

### Shortform Syntax Patterns

| Entity | Pattern | Example | Regex |
|--------|---------|---------|-------|
| **Program** | `Name -> EntryPoint [Purpose] [Version]` | `TodoApp -> AppEntry "Main application" v1.0.0` | `^(\w+)\s*->\s*(\w+)(?:\s+"([^"]+)")?(?:\s+v([\d.]+))?$` |
| **File** | `Name @ path:` | `UserService @ src/services/user.ts:` | `^(\w+)\s*@\s*([^:]+):` |
| **Function** | `Name :: Signature` | `createUser :: (data: UserDTO) => Promise<User>` | `^(\w+)\s*::\s*(.+)$` |
| **Class** | `Name <: BaseClass[, Interface1, Interface2]` | `UserController <: BaseController, IController` | `^(\w+)\s*<:\s*(.*)$` |
| **Class File** | `Name #: path [<: BaseClass[, Interface1, Interface2]]` | `UserController #: src/controllers/user.ts <: BaseController` | `^([A-Za-z][A-Za-z0-9_]*)\s*#:\s*([^\s<]+)(?:\s*<:\s*(.+))?$` |
| **Constants** | `Name ! path [: Schema]` | `Config ! src/config.ts : ConfigSchema` | `^(\w+)\s*!\s*([^:]+)(?:\s*:\s*(\w+))?$` |
| **Asset** | `Name ~ Description` | `Logo ~ "Company logo SVG"` | `^(\w+)\s*~\s*"([^"]+)"$` |
| **Ui Component** | `Name & Description | Name &! Description` | `App &! "Root application component"` | `^(\w+)\s*(&!?)\s*"([^"]+)"$` |
| **Run Parameter** | `Name $type Description [(required)]` | `DATABASE_URL $env "PostgreSQL connection" (required)` | `^(\w+)\s*\$(\w+)\s*"([^"]+)"(?:\s*\((\w+)\))?$` |
| **Dependency** | `Name ^ Purpose [Version]` | `axios ^ "HTTP client library" v3.0.0` | `^([@\w\-/]+)\s*\^\s*"([^"]+)"(?:\s*v?([\d.\-\w]+))?$` |

## Continuation Patterns

These patterns match continuation lines that add properties to entities:

| Pattern | Description | Example |
|---------|-------------|---------|
| Imports | Entity imports | `<- [Database, UserModel]` |
| Exports | Entity exports | `-> [createUser, getUser]` |
| Calls | Function calls | `~> [validate, save]` |
| Input | Function input DTO | `<- UserCreateDTO` |
| Output | Function output DTO | `-> UserDTO` |
| Methods | Class methods | `=> [create, read, update]` |
| Affects | Function affects UI | `~ [UserList, UserForm]` |
| Contains | UI component contains | `> [Header, Footer]` |
| Contained By | UI component parent | `< [Dashboard]` |
| Contains Program | Asset contains program | `>> ClientApp` |
| Dto Field | DTO field definition | `- name: string "User name"` |
| Comment | Comment line | `# This is a comment` |
| Description | Entity description | `"Creates a new user"` |
| Default Value | Parameter default | `= "default-value"` |
| Consumes | Function consumes params | `$< [DATABASE_URL, API_KEY]` |

## General Patterns

These patterns are used for general parsing tasks:

### Entity Declaration

**Description:** Detects any entity declaration line

**Regex:** `^[@\w\-/]+\s*(->|@|<:|#:|!|::|%|~|&|\$|\^|\s*:)`

### Longform Declaration

**Description:** Detects longform syntax declarations

**Regex:** `^(program|file|function|class|dto|component|asset|constants|parameter|import|dependency)\s+`

### Continuation

**Description:** Detects continuation lines for entity properties

**Regex:** `^\s+(->|<-|~>|=>|>>|>|<|~|"|#|-|=|\$<)`

### Import Statement

**Description:** Matches import statements (@import or import)

**Regex:** `^(?:@import|import)\s+"([^"]+)"(?:\s+as\s+(\w+))?$`

### Inline Comment

**Description:** Extracts inline comments from lines

**Regex:** `^(.+?)\s+#\s+(.+)$`

## Quick Reference Example

```tmd
TodoApp -> main v1.0.0                      # Program
main @ src/index.ts:                        # File
  <- [UserService]
  -> [startApp]

UserService #: src/services/user.ts         # ClassFile (fusion)
  <- [UserDTO]
  => [createUser, findUser]

startApp :: () => void                      # Function
  ~> [createUser]

createUser :: (data: UserDTO) => UserDTO    # Function
  <- UserDTO                                # Input DTO
  -> UserDTO                                # Output DTO

UserDTO %                                    # DTO
  - name: string "User name"
  - email: string "Email"

# Example showing other entity types:
App &! "Root component"                     # UIComponent (root)
DATABASE_URL $env "DB connection" (required) # RunParameter
Config ! src/config.ts                      # Constants
Logo ~ "Company logo"                       # Asset
react ^ "UI library" v18.0.0                # Dependency
```

## Key Features

### ClassFile Fusion (`#:`)
Combines Class and File into one entity - perfect for services/controllers:
```tmd
UserService #: src/services/user.ts <: BaseService
  <- [Database, Logger]       # File imports
  => [create, update, delete] # Class methods
  -> [userHelper]             # Additional exports
```

### Function Auto-Distribution
The `<- [...]` syntax intelligently categorizes mixed dependencies:
```tmd
processOrder :: (order: OrderDTO) => void
  <- [OrderDTO, validateOrder, Database, OrderUI, API_KEY]
  # Auto-distributed: input (DTO), calls (Functions/Classes),
  # affects (UI), consumes (RunParams/Assets/Constants)
```

## Validation Rules

### Bidirectional Consistency
TypedMind enforces bidirectional relationships:
- Function affects UIComponent → UIComponent.affectedBy includes Function
- Function consumes RunParameter → RunParameter.consumedBy includes Function
- UIComponent contains child → child.containedBy includes parent
- Asset contains Program → Program must exist

### Entity Naming Rules
- Names must be unique across ALL entity types
- Exception: ClassFile can replace separate Class + File with same name
- The validator will suggest using ClassFile fusion when detecting Class/File name conflicts

### Reference Type Validation
Each reference type has specific allowed source and target entity types:

| Reference | From Entities | To Entities |
|-----------|---------------|-------------|
| imports | File, Class, ClassFile | Function, Class, Constants, DTO, etc. |
| exports | File, ClassFile | Function, Class, Constants, DTO, etc. |
| calls | Function | Function, Class (for methods) |
| extends | Class, ClassFile | Class, ClassFile |
| affects | Function | UIComponent |
| consumes | Function | RunParameter, Asset, Constants |

## Parser Intelligence

### Context-Aware Parsing
The parser uses look-ahead to determine entity types:
- `Name @ path:` followed by `=> [methods]` → Class entity
- `Name @ path:` without methods → File entity
- Inline comments (`# comment`) are extracted and stored separately
- Mixed shortform/longform syntax is supported in the same file

### Import Resolution
- Circular imports are detected and prevented
- Aliased imports prefix all imported entities: `@import "./auth.tmd" as Auth`
- Nested imports are resolved recursively
- Import paths can be relative or absolute

## Operator Quick Reference

```
->  Entry point (Program) or Exports (File/Function)
<-  Imports or Dependencies
@   File path location
#:  ClassFile fusion (class + file)
::  Function signature
<:  Class inheritance
!   Constants marker
%   DTO marker
~   Asset description or Function affects UI
&   UIComponent (&! for root)
$   RunParameter ($env, $iam, etc.)
^   External dependency
~>  Function calls
=>  Class methods
>>  Asset contains program
>   UIComponent contains
<   UIComponent contained by
$<  Function consumes parameters
:   Constants schema
=   Parameter default value
```

### DTOs vs Classes
**DTOs**: Pure data structures (NO functions allowed)
**Classes**: Behavior and business logic (have methods)
```tmd
UserDTO %                            # DTO: data only
  - name: string "User name"
  - email: string

UserService #: src/services/user.ts # Class: behavior
  => [createUser, findUser]         # Has methods
```

## Advanced Patterns via Purpose Fields

The purpose field can capture advanced programming patterns that TypedMind structure alone cannot represent:

### Pattern Examples
```tmd
# Example patterns - showing syntax only (not complete programs)
# Async/Concurrent
processWorker :: (jobs: Channel<Job>) => void "ASYNC: Goroutine worker"
DataChannel % "CHANNEL: MPSC unbounded"

# Generics/Templates
Container<T> <: Base "GENERIC<T: Display>: Type-parameterized"

# Dependency Injection
UserService #: src/service.ts "@Injectable @Scope(singleton)"

# Event-Driven
Button & "Component" "EVENTS: onClick, onHover, onFocus"
DataEmitter <: EventEmitter "EMITS: data, error, close"

# Resource Management
FileReader :: (path: string) => string "RAII: auto-closes handle"
Connection % "Context manager: auto-commit on scope exit"

# Build Configuration
DebugLogger ! src/debug.ts "BUILD: #ifdef DEBUG only"

# Pattern Matching
handleOption :: (val: Option<T>) => string "MATCH: Some(x) | None"

# Middleware/Pipeline
AuthMiddleware :: (req, res, next) => void "MIDDLEWARE: JWT validation"
Pipeline @ src/pipeline.ts: "PIPELINE: cors -> auth -> router"
```

### Semantic Conventions
Establish project-specific conventions in purpose fields:
- **ASYNC/AWAIT**: Async functions and promises
- **GENERIC<T>**: Generic type parameters
- **@Decorator**: Decorators and annotations
- **EVENTS**: Event emitters and handlers
- **CHANNEL**: Concurrent communication
- **RAII/Context**: Resource management
- **BUILD**: Conditional compilation
- **PIPELINE**: Middleware chains

## Best Practices

- **Use ClassFile (`#:`)** for services, controllers, repositories
- **Group by feature**: Keep related entities together
- **Mix dependencies freely**: Parser auto-categorizes them
- **DTOs for data, Classes for behavior**: Keep them separate
- **Leverage purpose fields**: Document async, generics, DI, events, etc.
- **Establish conventions**: Create project-specific semantic patterns
- **Bidirectional links**: Validator ensures consistency