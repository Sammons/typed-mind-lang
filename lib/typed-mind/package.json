{
  "name": "@sammons/typed-mind",
  "version": "1.0.0",
  "description": "Core language parser and validator for TypedMind DSL",
  "author": "sammons",
  "license": "MIT",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsc --build",
    "dev": "tsc --build --watch",
    "clean": "rm -rf dist *.tsbuildinfo",
    "test": "vitest --run",
    "test:watch": "vitest --watch",
    "lint": "eslint src"
  },
  "devDependencies": {
    "@types/node": "^22.10.5",
    "tsup": "^8.3.5",
    "vitest": "^2.1.8"
  },
  "publishConfig": {
    "access": "public"
  }
}