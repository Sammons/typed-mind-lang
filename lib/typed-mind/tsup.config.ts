import { defineConfig } from 'tsup';

export default defineConfig({
  entry: ['src/index.ts'],
  format: ['cjs'],
  dts: {
    resolve: true,
  },
  clean: true,
  sourcemap: true,
  splitting: false,
  treeshake: true,
  tsconfig: './tsconfig.json',
});