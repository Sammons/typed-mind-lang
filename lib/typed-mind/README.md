# @sammons/typed-mind

Core language parser and validator for TypedMind DSL - a declarative language for building software architectures.

## Installation

```bash
npm install @sammons/typed-mind
```

## Usage

```typescript
import { TypedMindParser, TypedMindValidator } from '@sammons/typed-mind';

// Parse TypedMind source code
const parser = new TypedMindParser();
const program = parser.parse(source);

// Validate the parsed program
const validator = new TypedMindValidator();
const validationResult = validator.validate(program);

if (validationResult.isValid) {
  console.log('Program is valid!');
} else {
  console.log('Validation errors:', validationResult.errors);
}
```

## What is TypedMind?

TypedMind is a Domain Specific Language (DSL) for declaratively describing software architectures, including:

- **Data Transfer Objects (DTOs)** - Define data structures and their validation rules
- **Service classes** - Specify business logic and their dependencies
- **UI components** - Describe user interface elements and their relationships
- **Asset management** - Define static assets and routing configuration
- **Cross-cutting concerns** - Handle validation, security, and other system-wide concerns

## Features

- **Declarative syntax** - Focus on what your system does, not how it's implemented
- **Strong validation** - Catch architectural issues before they become problems
- **Dependency tracking** - Understand relationships between components
- **Import system** - Modular architecture with file-based organization
- **Type safety** - Comprehensive validation of all language constructs

## Example

```typedmind
program "User Management System"

dto UserDto {
  id: string
  email: string
  name: string
  createdAt: date
}

service UserService {
  method createUser(email: string, name: string): UserDto
  method getUserById(id: string): UserDto
  depends on DatabaseService
}

ui UserListComponent {
  displays UserDto[]
  triggers UserService.createUser
}
```

## Grammar

The complete TypedMind grammar is documented in [grammar.md](./grammar.md) and available in EBNF format in [grammar.ebnf](./grammar.ebnf).

## Requirements

- Node.js >= 22.0.0

## Related Packages

- [@sammons/typed-mind-cli](https://www.npmjs.com/package/@sammons/typed-mind-cli) - Command-line interface
- [@sammons/typed-mind-lsp](https://www.npmjs.com/package/@sammons/typed-mind-lsp) - Language Server Protocol implementation
- [@sammons/typed-mind-renderer](https://www.npmjs.com/package/@sammons/typed-mind-renderer) - HTML/D3.js renderer

## License

MIT