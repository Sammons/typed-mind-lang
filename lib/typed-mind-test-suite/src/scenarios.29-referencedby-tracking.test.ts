import { describe, it, expect } from 'vitest';
import { readFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { DSLParser } from '@sammons/typed-mind';
import { DSLValidator } from '@sammons/typed-mind';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

describe('scenario-29-referencedby-tracking', () => {
  const scenarioFile = 'scenario-29-referencedby-tracking.tmd';

  it('should track ReferencedBy relationships', () => {
    const filePath = join(__dirname, '..', 'scenarios', scenarioFile);
    const content = readFileSync(filePath, 'utf-8');
    
    const parser = new DSLParser();
    const validator = new DSLValidator();
    
    const parseResult = parser.parse(content);
    const validationResult = validator.validate(parseResult.entities);
    
    if (!validationResult.valid) {
      console.log('Validation errors:', validationResult.errors);
    }
    
    expect(validationResult.valid).toBe(true);
    
    // Test File references
    const userService = parseResult.entities.get('UserService');
    expect(userService).toBeDefined();
    expect(userService?.referencedBy?.some(ref => ref.from === 'MainFile' && ref.type === 'imports')).toBe(true);
    
    // Test Function references
    const createUser = parseResult.entities.get('createUser');
    expect(createUser).toBeDefined();
    expect(createUser?.referencedBy?.some(ref => ref.from === 'UserService' && ref.type === 'exports')).toBe(true);
    
    // Test DTO references
    const userDTO = parseResult.entities.get('UserDTO');
    expect(userDTO).toBeDefined();
    expect(userDTO?.referencedBy?.some(ref => ref.from === 'MainFile' && ref.type === 'imports')).toBe(true);
    expect(userDTO?.referencedBy?.some(ref => ref.from === 'createUser' && ref.type === 'input')).toBe(true);
    
    const user = parseResult.entities.get('User');
    expect(user).toBeDefined();
    expect(user?.referencedBy?.some(ref => ref.from === 'createUser' && ref.type === 'output')).toBe(true);
    expect(user?.referencedBy?.some(ref => ref.from === 'getUser' && ref.type === 'output')).toBe(true);
    
    // Test Class references
    const database = parseResult.entities.get('Database');
    expect(database).toBeDefined();
    expect(database?.referencedBy?.some(ref => ref.from === 'UserService' && ref.type === 'imports')).toBe(true);
    expect(database?.referencedBy?.some(ref => ref.from === 'createUser' && ref.type === 'calls')).toBe(true);
    expect(database?.referencedBy?.some(ref => ref.from === 'getUser' && ref.type === 'calls')).toBe(true);
    
    // Test Constants references
    const databaseConfig = parseResult.entities.get('DatabaseConfig');
    expect(databaseConfig).toBeDefined();
    expect(databaseConfig?.referencedBy?.some(ref => ref.from === 'MainFile' && ref.type === 'imports')).toBe(true);
    
    // Test Program entry references
    const mainFile = parseResult.entities.get('MainFile');
    expect(mainFile).toBeDefined();
    expect(mainFile?.referencedBy?.some(ref => ref.from === 'TestApp' && ref.type === 'entry')).toBe(true);
    
    // Test UIComponent references
    const userList = parseResult.entities.get('UserList');
    expect(userList).toBeDefined();
    expect(userList?.referencedBy?.some(ref => ref.from === 'AppUI' && ref.type === 'contains')).toBe(true);
    
    const appUI = parseResult.entities.get('AppUI');
    expect(appUI).toBeDefined();
    expect(appUI?.referencedBy?.some(ref => ref.from === 'UserList' && ref.type === 'containedBy')).toBe(true);
    expect(appUI?.referencedBy?.some(ref => ref.from === 'UserForm' && ref.type === 'containedBy')).toBe(true);
    
    // Test RunParameter references
    const databaseUrl = parseResult.entities.get('DATABASE_URL');
    expect(databaseUrl).toBeDefined();
    expect(databaseUrl?.referencedBy?.some(ref => ref.from === 'handler' && ref.type === 'consumes')).toBe(true);
    
    // Test Asset program references
    const clientProgram = parseResult.entities.get('ClientProgram');
    expect(clientProgram).toBeDefined();
    expect(clientProgram?.referencedBy?.some(ref => ref.from === 'HTMLAsset' && ref.type === 'containsProgram')).toBe(true);
    
    // Create snapshot for validation
    const output = {
      file: scenarioFile,
      valid: validationResult.valid,
      errors: validationResult.errors.map(err => ({
        line: err.position.line,
        column: err.position.column,
        message: err.message,
        severity: err.severity,
        suggestion: err.suggestion
      })),
      // Include some key referencedBy relationships in snapshot
      referencedByExamples: {
        UserService: userService?.referencedBy,
        UserDTO: userDTO?.referencedBy,
        Database: database?.referencedBy,
        UserList: userList?.referencedBy,
        DATABASE_URL: databaseUrl?.referencedBy,
        ClientProgram: clientProgram?.referencedBy
      }
    };
    
    expect(output).toMatchSnapshot();
  });
});