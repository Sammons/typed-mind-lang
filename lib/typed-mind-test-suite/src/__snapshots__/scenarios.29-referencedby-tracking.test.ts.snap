// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`scenario-29-referencedby-tracking > should track ReferencedBy relationships 1`] = `
{
  "errors": [],
  "file": "scenario-29-referencedby-tracking.tmd",
  "referencedByExamples": {
    "ClientProgram": [
      {
        "from": "HTMLAsset",
        "fromType": "Asset",
        "type": "containsProgram",
      },
    ],
    "DATABASE_URL": [
      {
        "from": "handler",
        "fromType": "Function",
        "type": "consumes",
      },
    ],
    "Database": [
      {
        "from": "UserService",
        "fromType": "File",
        "type": "imports",
      },
      {
        "from": "UserService",
        "fromType": "File",
        "type": "exports",
      },
      {
        "from": "createUser",
        "fromType": "Function",
        "type": "calls",
      },
      {
        "from": "getUser",
        "fromType": "Function",
        "type": "calls",
      },
    ],
    "UserDTO": [
      {
        "from": "MainFile",
        "fromType": "File",
        "type": "imports",
      },
      {
        "from": "createUser",
        "fromType": "Function",
        "type": "input",
      },
    ],
    "UserList": [
      {
        "from": "AppUI",
        "fromType": "UIComponent",
        "type": "contains",
      },
    ],
    "UserService": [
      {
        "from": "MainFile",
        "fromType": "File",
        "type": "imports",
      },
    ],
  },
  "valid": true,
}
`;
